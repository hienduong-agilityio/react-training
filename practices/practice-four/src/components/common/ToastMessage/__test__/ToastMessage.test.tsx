// Libraries
import { render, screen, act } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

// Components
import Toast from '@/components/common/ToastMessage';

/*
Toast Component
  √ should match snapshot
  √ should render success toast with default content
  √ should render error toast with custom content
  √ should render with undo option when undoEnabled is true
  √ should call onClose after timeoutDuration
*/
describe('Toast Component', () => {
  const onCloseMock = jest.fn();
  let renderComponent: (props?: Partial<React.ComponentProps<typeof Toast>>) => ReturnType<typeof render>;

  beforeEach(() => {
    jest.clearAllMocks();
    renderComponent = (props = {}) =>
      render(
        <MemoryRouter>
          <Toast onClose={onCloseMock} {...props}>
            {props.children || 'Default toast message'}
          </Toast>
        </MemoryRouter>
      );
  });

  it('should match snapshot', () => {
    const { container } = renderComponent({ children: 'Snapshot test' });
    expect(container).toMatchSnapshot();
  });

  it('should render success toast with default content', () => {
    const { container } = renderComponent({ children: 'This is a success message' });

    const toastElement = screen.getByText('This is a success message');
    expect(toastElement).toBeInTheDocument();
    expect(container.firstChild).toHaveClass('text-success-100 bg-success-800');
  });

  it('should render error toast with custom content', () => {
    const { container } = renderComponent({ type: 'error', children: 'Error occurred' });

    const toastElement = screen.getByText('Error occurred');
    expect(toastElement).toBeInTheDocument();
    expect(container.firstChild).toHaveClass('text-red-100 bg-red-800');
  });

  it('should render with undo option when undoEnabled is true', () => {
    renderComponent({ undoEnabled: true, children: 'Undo test' });

    const undoLink = screen.getByText('Undo');
    expect(undoLink).toBeInTheDocument();
    expect(undoLink).toHaveAttribute('href', '/projects');
  });

  it('should call onClose after timeoutDuration', async () => {
    jest.useFakeTimers();

    act(() => {
      jest.advanceTimersByTime(3000);
    });

    expect(onCloseMock).toHaveBeenCalledTimes(0);
    jest.useRealTimers();
  });
});
