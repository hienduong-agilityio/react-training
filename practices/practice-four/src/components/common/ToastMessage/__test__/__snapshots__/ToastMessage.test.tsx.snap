// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast Component should match snapshot 1`] = `
<div>
  <div
    class="flex items-center divide-x rtl:divide-x-reverse p-4 w-full max-w-xs rounded-lg shadow text-success-100 bg-success-800 divide-green-950"
  >
    <div
      class="text-sm font-normal inline-flex items-center gap-3 pl-1"
    >
      <img
        alt="success"
      />
      Snapshot test
    </div>
  </div>
</div>
`;
